README.md
# About

slurm_inspector V0.1 (2015.12.20), written by Willi Kappler

Licensed under the MIT License

This tool monitors the SLURM job management system (SLURM = Simple Linux Utility for Resource Management)
(see [https://computing.llnl.gov/linux/slurm/slurm.html](https://computing.llnl.gov/linux/slurm/slurm.html))
The result is presented on a simple web page (using default values for ex. http://localhost:4545).
slurm_inspector is completely written in [Rust](https://www.rust-lang.org/) using the [Iron](http://ironframework.io/) web framework.

# Install

1. Obtain source code from github
2. Compile and run it with
    cargo run --release

Since Iron is based on OpenSSL You may have to install the OpenSSL development package first.

For ubuntu:

    sudo apt-get install libssl-dev

# Usage

slurm_inspector offers the following command line options:

    -p --port=[PORT] Sets the port for the web GUI (default: 4545)

    -i --interval=[INTERVAL] Sets the update interval (in sec.) for the web page (default: 60 sec.)

    --test create test values, does not call sinfo or squeue

    --loglevel=[LOGLEVEL] specify log level: error, info or debug

For example:

    cargo run --release -- -p 1234 -i 120

will run slurm_inspector listening on port 1234 and refreshing the SLURM status every 120 seconds.
Start your web browser and go to http://localhost:1234 (or http://myserver.com:1234)

The file "slurm_inspector.conf" contains an example ubuntu service configuration (and some comments on how to install it)

# TODO
- Use https instead of http
- Use login mechanism (user name and password)
- Use HTML template mechanism (HandleBars, Roustache, ...)
- Show more SLURM information values (accounting, etc.)
- Add some fancy charts and shiny CSS
- Maybe some interaction with JavaScript
- Clean up code (better error handling, ...)
- More test cases
- Better documentation
- Maybe use libslurm instead of parsing output of "sinfo" and "squeue"

Any feedback is welcome! If you have any ideas or an important feature is missing or if you found a bug let me know.


CHANGELOG.md
# Change Log

## 0.1 - 2015-12-20

First public release on github


tests/lib.rs
extern crate slurm_util;

// use slurm_util::sinfo_util::*;

// TODO: add some integration tests
#[test]
fn test_lib() {
    assert!(true);
}


src/main.rs
// slurm_inspector V0.1 (2015.12.20), written by Willi Kappler
//
// Licensed under the MIT License

// External crates:
extern crate iron;
#[macro_use] extern crate log;
extern crate flexi_logger;

// Internal crates:
extern crate slurm_util;

// System modules:
use std::net::{SocketAddrV4, Ipv4Addr};
use std::sync::{Arc, Mutex};

// External modules:
use iron::prelude::{Iron, Request};
use flexi_logger::{detailed_format,init,LogConfig};

// Internal modules:
use slurm_util::configuration::setup_configuration;
use slurm_util::request_handler::handle_request;
use slurm_util::slurm_status::{SlurmStatus, check_slurm_status};

fn main() {
    // Parse command line arguments
    let config = setup_configuration();

    init(LogConfig { log_to_file: true, format: detailed_format, .. LogConfig::new() }, Some(config.log_level.clone()))
         .unwrap_or_else(|e| { panic!("Logger initialization failed with the following error: {}", e) });

    info!("configuration: port: {}, interval: {}, test mode: {}, log level: {}", config.port, config.interval, config.test_mode, config.log_level);

    // Create empty SlurmStatus object
    let initial_slurm_status = SlurmStatus{ node_info: Vec::new(), job_info: Vec::new(), last_update: String::new() };

    // Iron-persistence can't be used here since own thread can't access private filed "data" of struct "State"
    let local_slurm_status = Arc::new(Mutex::new(initial_slurm_status));

    // Start the background thread and read in real values
    check_slurm_status(&local_slurm_status, config.clone());

    // Need to clone this since each client request will be handled by iron in a separate thread
    let shared_slurm_status = local_slurm_status.clone();

    // Run iron web framework and wait for the user to load the page
    Iron::new( move |req: &mut Request| { handle_request(req, &shared_slurm_status) }
).http(SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), config.port)).unwrap();
}


src/sinfo_util.rs
//! Abstraction for the SLURM sinfo command
//! Runs sinfo, parses output into data structure (PartitionNodeInfo)

// System modules:
use std::process::Command;

/// PartitionAvailability, can be "up" or "down"
#[derive(Debug, PartialEq)]
pub enum PartitionAvailability {
    /// SLURM Partition is up and available
    Up,
    /// SLURM partition is not available
    Down
}

/// The ErrorCause why a SLURM partition is down
#[derive(Debug, PartialEq)]
pub enum ErrorCause {
    Down,
    Drained,
    Draining,
    None,
    /// An unknown error occured
    Unknown
}

/// The state of an individual node
#[derive(Debug, PartialEq)]
pub enum NodeState {
    Allocated,
    Completing,
    Down,
    Drained,
    Draining,
    Fail,
    Failing,
    Idle,
    Maint,
    /// The node is in an unknown state
    Unknown
}

/// SLURM partition and node information
#[derive(Debug, PartialEq)]
pub struct PartitionNodeInfo {
    pub partition: String,
    pub availability: PartitionAvailability,
    pub hostname: String,
    pub node: String,
    pub error: ErrorCause,
    pub cpu_load: Option<f64>,
    pub node_state: NodeState,
    pub node_sockets: Option<u32>,
    pub node_cores: Option<u32>,
    pub node_threads: Option<u32>
}

/// Public helper function to generate test data
pub fn get_partition_node_info_test() -> Vec<PartitionNodeInfo> {
    let test_data = "
        esd up node01 node01 none 0.22 idle 2 2 2
        esd up node02 node02 none 0.0 idle 2 8 2
        esd down node03 node03 down - - - - -
        esd up node04 node04 none 0.71 idle 1 1 1
        esd up node05 node05 none 0.0 alloc 1 1 1
        esd up node06 node06 none 0.0 completing 1 1 1
        esd up node07 node07 none 0.0 drained 1 1 1
        esd up node08 node08 none 0.0 draining 1 1 1
        esd up node09 node09 none 0.0 fail 1 1 1
        esd up node10 node10 none 0.0 failing 1 1 1
        esd up node11 node11 none 0.0 maint 1 1 1
        esd up node12 node12 none 0.0 unknown 1 1 1
    ";

    get_pn_info_util(test_data)
}

/// Public helper function to retrieve a list of current SLURM partition and node status
pub fn get_partition_node_info() -> Vec<PartitionNodeInfo> {
    get_pn_info_util(&call_sinfo())
}

// Private helper function to parse the output of "sinfo" and return a list of PartitionNodeInfo
fn get_pn_info_util(sinfo_output: &str) -> Vec<PartitionNodeInfo> {
    let mut result: Vec<PartitionNodeInfo> = Vec::new();

    for line in sinfo_output.lines() {
        let items: Vec<&str> = line.split_whitespace().collect();

        // Skip invalid lines
        if items.len() != 10 {
            debug!("number of items in line: {}", items.len());
            continue;
        }

        result.push( PartitionNodeInfo{
            partition: items[0].to_string(),
            availability: str_to_availability(items[1]),
            hostname: items[2].to_string(),
            node: items[3].to_string(),
            error: str_to_error(items[4]),
            cpu_load: items[5].parse::<f64>().ok(),
            node_state: str_to_node_state(items[6]),
            node_sockets: items[7].parse::<u32>().ok(),
            node_cores: items[8].parse::<u32>().ok(),
            node_threads: items[9].parse::<u32>().ok()
        })
    }

    result
}

#[test]
fn test_get_pn_info_util_empty() {
    assert!(get_pn_info_util("").len() == 0);
}

#[test]
fn test_get_pn_info_util_invalid() {
    assert!(get_pn_info_util("1 2 3").len() == 0);
}

#[test]
fn test_get_pn_info_util_01() {
    let input = "longrun up node01.foo.bar node01 none 0.22 idle 2 2 2";
    let output = vec![PartitionNodeInfo{
        partition: "longrun".to_string(),
        availability: PartitionAvailability::Up,
        hostname: "node01.foo.bar".to_string(),
        node: "node01".to_string(),
        error: ErrorCause::None,
        cpu_load: Some(0.22),
        node_state: NodeState::Idle,
        node_sockets: Some(2),
        node_cores: Some(2),
        node_threads: Some(2)
    }];

    assert_eq!(get_pn_info_util(input), output);
}

#[test]
fn test_get_pn_info_util_02() {
    let input = "longrun up node01.foo.bar node01 none 0.22 idle 2 2 2\nlongrun up node02.foo.bar node02 down 0.1 idle 1 2 4";
    let output = vec![
        PartitionNodeInfo{
            partition: "longrun".to_string(),
            availability: PartitionAvailability::Up,
            hostname: "node01.foo.bar".to_string(),
            node: "node01".to_string(),
            error: ErrorCause::None,
            cpu_load: Some(0.22),
            node_state: NodeState::Idle,
            node_sockets: Some(2),
            node_cores: Some(2),
            node_threads: Some(2)
        },
        PartitionNodeInfo{
            partition: "longrun".to_string(),
            availability: PartitionAvailability::Up,
            hostname: "node02.foo.bar".to_string(),
            node: "node02".to_string(),
            error: ErrorCause::Down,
            cpu_load: Some(0.1),
            node_state: NodeState::Idle,
            node_sockets: Some(1),
            node_cores: Some(2),
            node_threads: Some(4)
        }
    ];

    assert_eq!(get_pn_info_util(input), output);
}

/*
    sinfo -o "%R %a %n %N %E %O %T %X %Y %Z" -h
    %R: partition name
    %a: availability of partition (up / down)
    %n: host name
    %N: node name
    %E: reason of error if any (down, drained, draining, none)
    %O: CPU load of node
    %T: sdtate of node
    %X: number of sockets per node
    %Y: number of cores per node
    %Z: number of threads per node

    example output:

    high_mem up node01.foo.bar node01 none 0.01 idle 2 2 2
    high_mem up node02.foo.bar node02 none 0.02 idle 2 2 2
    high_mem up node03.foo.bar node03 none 0.12 idle 2 2 2
    high_mem up node04.foo.bar node04 none 0.03 idle 2 6 2
    high_mem up node05.foo.bar node05 none 0.01 idle 2 4 2
    high_mem up node06.foo.bar node06 none 0.01 idle 2 4 2
    high_mem up node07.foo.bar node07 none 0.01 idle 2 4 2
    high_mem up node08.foo.bar node08 none 0.01 idle 2 8 2
    high_mem up node09.foo.bar node09 none 0.01 idle 2 8 2
    high_mem up node10.foo.bar node10 none 0.01 idle 2 8 2
    high_mem up node11.foo.bar node11 none 0.01 idle 2 2 2
*/

// Private helper function to execute the "sinfo" SLURM command and return its output as a string
// On error returns an empty string. TODO: better error handling
fn call_sinfo() -> String {
    let output = Command::new("sinfo")
        .arg("-h")
        .arg("-o")
        .arg("%R %a %n %N %E %O %T %X %Y %Z")
        .output();

    match output {
        Result::Ok(val) => String::from_utf8_lossy(&val.stdout).to_string(),
        Result::Err(err) => {
            error!("Could not execute 'sinfo': {}", err);
            // return empty string on error, but continue with the program
            String::new()
        }
    }
}

// Private helper function to parse partition availability
fn str_to_availability(avail: &str) -> PartitionAvailability {
    match &*avail.to_lowercase() {
        "up" => PartitionAvailability::Up,
        _ => PartitionAvailability::Down
    }
}

#[test]
fn test_str_to_availability_up() {
    assert_eq!(str_to_availability("up"), PartitionAvailability::Up);
    assert_eq!(str_to_availability("Up"), PartitionAvailability::Up);
    assert_eq!(str_to_availability("UP"), PartitionAvailability::Up);
}

#[test]
fn test_str_to_availability_down() {
    assert_eq!(str_to_availability("down"), PartitionAvailability::Down);
    assert_eq!(str_to_availability("Down"), PartitionAvailability::Down);
    assert_eq!(str_to_availability("DOWN"), PartitionAvailability::Down);
}

// Private helper function to parse partition error cause
fn str_to_error(error: &str) -> ErrorCause {
    match &*error.to_lowercase() {
        "down" => ErrorCause::Down,
        "drained" => ErrorCause::Drained,
        "draining" => ErrorCause::Draining,
        "none" => ErrorCause::None,
        _ => ErrorCause::Unknown
    }
}

#[test]
fn test_str_to_error_down() {
    assert_eq!(str_to_error("down"), ErrorCause::Down);
    assert_eq!(str_to_error("Down"), ErrorCause::Down);
    assert_eq!(str_to_error("DOWN"), ErrorCause::Down);
}

#[test]
fn test_str_to_error_drained() {
    assert_eq!(str_to_error("drained"), ErrorCause::Drained);
    assert_eq!(str_to_error("Drained"), ErrorCause::Drained);
    assert_eq!(str_to_error("DRAINED"), ErrorCause::Drained);
}

#[test]
fn test_str_to_error_draining() {
    assert_eq!(str_to_error("draining"), ErrorCause::Draining);
    assert_eq!(str_to_error("Draining"), ErrorCause::Draining);
    assert_eq!(str_to_error("DRAINING"), ErrorCause::Draining);
}

#[test]
fn test_str_to_error_none() {
    assert_eq!(str_to_error("none"), ErrorCause::None);
    assert_eq!(str_to_error("None"), ErrorCause::None);
    assert_eq!(str_to_error("NONE"), ErrorCause::None);
}

// Private helper function to parse node state
fn str_to_node_state(n_state: &str) -> NodeState {
    match &*n_state.to_lowercase() {
            "alloc" | "allocated" => NodeState::Allocated,
            "completing" => NodeState::Completing,
            "down" => NodeState::Down,
            "drained" => NodeState::Drained,
            "draining" => NodeState::Draining,
            "fail" => NodeState::Fail,
            "failing" => NodeState::Failing,
            "idle" => NodeState::Idle,
            "maint" => NodeState::Maint,
            _ => NodeState::Unknown
    }
}

#[test]
fn test_str_to_node_state_alloc() {
    assert_eq!(str_to_node_state("alloc"), NodeState::Allocated);
    assert_eq!(str_to_node_state("Alloc"), NodeState::Allocated);
    assert_eq!(str_to_node_state("ALLOC"), NodeState::Allocated);
    assert_eq!(str_to_node_state("allocated"), NodeState::Allocated);
    assert_eq!(str_to_node_state("allocated"), NodeState::Allocated);
    assert_eq!(str_to_node_state("ALLOCATED"), NodeState::Allocated);
}

#[test]
fn test_str_to_node_state_completing() {
    assert_eq!(str_to_node_state("completing"), NodeState::Completing);
    assert_eq!(str_to_node_state("Completing"), NodeState::Completing);
    assert_eq!(str_to_node_state("COMPLETING"), NodeState::Completing);
}

#[test]
fn test_str_to_node_state_down() {
    assert_eq!(str_to_node_state("down"), NodeState::Down);
    assert_eq!(str_to_node_state("Down"), NodeState::Down);
    assert_eq!(str_to_node_state("DOWN"), NodeState::Down);
}

#[test]
fn test_str_to_node_state_drained() {
    assert_eq!(str_to_node_state("drained"), NodeState::Drained);
    assert_eq!(str_to_node_state("Drained"), NodeState::Drained);
    assert_eq!(str_to_node_state("DRAINED"), NodeState::Drained);
}

#[test]
fn test_str_to_node_state_draining() {
    assert_eq!(str_to_node_state("draining"), NodeState::Draining);
    assert_eq!(str_to_node_state("Draining"), NodeState::Draining);
    assert_eq!(str_to_node_state("DRAINING"), NodeState::Draining);
}

#[test]
fn test_str_to_node_state_fail() {
    assert_eq!(str_to_node_state("fail"), NodeState::Fail);
    assert_eq!(str_to_node_state("Fail"), NodeState::Fail);
    assert_eq!(str_to_node_state("FAIL"), NodeState::Fail);
}

#[test]
fn test_str_to_node_state_failing() {
    assert_eq!(str_to_node_state("failing"), NodeState::Failing);
    assert_eq!(str_to_node_state("Failing"), NodeState::Failing);
    assert_eq!(str_to_node_state("FAILING"), NodeState::Failing);
}

#[test]
fn test_str_to_node_state_idle() {
    assert_eq!(str_to_node_state("idle"), NodeState::Idle);
    assert_eq!(str_to_node_state("Idle"), NodeState::Idle);
    assert_eq!(str_to_node_state("IDLE"), NodeState::Idle);
}

#[test]
fn test_str_to_node_state_maint() {
    assert_eq!(str_to_node_state("maint"), NodeState::Maint);
    assert_eq!(str_to_node_state("Maint"), NodeState::Maint);
    assert_eq!(str_to_node_state("MAINT"), NodeState::Maint);
}

#[test]
fn test_str_to_node_state_unknown() {
    assert_eq!(str_to_node_state("unknown"), NodeState::Unknown);
    assert_eq!(str_to_node_state("Unknown"), NodeState::Unknown);
    assert_eq!(str_to_node_state("UNKNOWN"), NodeState::Unknown);
}


src/slurm_status.rs
//! SLURM status data structure, updating and converting to HTML
//! Contains partition, node and job information and also time of last update
//! Runs a background thread in an endless loop to periodically check SLURM status
//! and updates data structure accordingly

// System modules:
use std::sync::{Arc, Mutex};
use std::thread;
use std::thread::sleep;
use std::time::Duration;

// External modules:
use time::{strftime, now};

// Internal modules:
use sinfo_util::{PartitionNodeInfo, PartitionAvailability, get_partition_node_info, get_partition_node_info_test};
use squeue_util::{JobInfo, get_job_info, get_job_info_test};
use configuration::Configuration;

/// SlurmStatus type. This is shared accross the Iron response threads and the SLURM status thread
pub struct SlurmStatus {
    /// List of partition and node information
    pub node_info: Vec<PartitionNodeInfo>,
    /// List of job information
    pub job_info: Vec<JobInfo>,
    /// The last time the above two lists have been updated
    /// Format: %Y.%m.%d - %H:%M
    pub last_update: String
}

/// Public function that starts the SLURM status thread and update the SlurmStatus object accordingly every time interval
/// TODO: better error handling
pub fn check_slurm_status(local_slurm_status: &Arc<Mutex<SlurmStatus>>, config: Configuration) {
    let shared_slurm_status = local_slurm_status.clone();

    thread::spawn(move || {
        // Endless loop, just keep checking the status of SLURM
        loop {
            match shared_slurm_status.lock() {
                Ok(mut status) => {
                    debug!("Update slurm status");
                    if config.test_mode {
                        status.node_info = get_partition_node_info_test();
                        status.job_info = get_job_info_test();
                    } else {
                        status.node_info = get_partition_node_info();
                        status.job_info = get_job_info();
                    }
                    status.last_update = strftime("%Y.%m.%d - %H:%M", &now()).unwrap();
                },
                Err(err) => {
                    error!("Could not lock Mutex: {}", err);
                }
            }

            sleep(Duration::new(config.interval, 0));
        }
    });
}

/// Public helper function accepts SlurmStatus and returns a string containing the HTML representation of the status
/// TODO: use some template mechanism (HandleBars, Roustache, ...)
pub fn status_to_html(status: &SlurmStatus) -> String {
    let mut result = String::new();

    // Page header and some CSS
    result.push_str("<html>\n");
    result.push_str("<head>\n");
    result.push_str("<title>Slurm Inspector</title>\n");
    result.push_str("<style>\n");
    result.push_str("table, { border: 1px solid black; }\n");
    result.push_str("th, td { border: 1px solid black; padding: 10px; }\n");
    result.push_str("th { background: #e0e0e0; }\n");
    result.push_str("#partition_down { background: #ffa0a0; }\n");
    result.push_str("</style>\n");
    result.push_str("</head>\n");
    result.push_str("<body>\n");

    // When was the SLURM status information last updated ?
    result.push_str(&format!("<h3>Last update: {}</h3>", status.last_update));

    result.push_str("<br>\n<br>\n<br>\n<br>\n");

    // Prepare first table (partition and node) with header
    result.push_str("<h3>Partition and node information:</h3>\n");
    result.push_str("<table>\n");
    result.push_str("<tr>\n");
    result.push_str("<th>Partition</th>");
    result.push_str("<th>Availability</th>");
    result.push_str("<th>Hostname</th>");
    result.push_str("<th>Node</th>");
    result.push_str("<th>Error</th>");
    result.push_str("<th>CPU load</th>");
    result.push_str("<th>Node state</th>");
    result.push_str("<th>Node sockets</th>");
    result.push_str("<th>Node cores</th>");
    result.push_str("<th>Node threads</th>");
    result.push_str("</tr>\n");

    for node in &status.node_info {
        result.push_str("<tr>\n");
        result.push_str(&format!("<td>{}</td>", node.partition));
        result.push_str(
            match node.availability {
                PartitionAvailability::Up => "<td>Up</td>",
                _ => "<td id=\"partition_down\">Down</td>"
            }
        );
        result.push_str(&format!("<td>{}</td>", node.hostname));
        result.push_str(&format!("<td>{}</td>", node.node));
        result.push_str(&format!("<td>{:?}</td>", node.error));
        result.push_str(node.cpu_load.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(format!("<td>{:?}</td>", node.node_state).as_ref());
        result.push_str(node.node_sockets.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(node.node_cores.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(node.node_threads.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str("</tr>\n");
    }

    result.push_str("</table>\n");

    result.push_str("<br>\n<br>\n<br>\n<br>\n");

    // Prepare second table (job information) with header
    result.push_str("<h3>Job information:</h3>\n");
    result.push_str("<table>\n");
    result.push_str("<tr>\n");
    result.push_str("<th>Executing host</th>");
    result.push_str("<th>Min CPU</th>");
    result.push_str("<th>Num CPU</th>");
    result.push_str("<th>Num nodes</th>");
    result.push_str("<th>Job array ID</th>");
    result.push_str("<th>Number of Sockets</th>");
    result.push_str("<th>Job ID</th>");
    result.push_str("<th>Number of Cores</th>");
    result.push_str("<th>Job name</th>");
    result.push_str("<th>Number of threads</th>");
    result.push_str("<th>Job array index</th>");
    result.push_str("<th>Run time</th>");
    result.push_str("<th>List of nodes</th>");
    result.push_str("<th>Priority</th>");
    result.push_str("<th>State reason</th>");
    result.push_str("<th>Start time</th>");
    result.push_str("<th>Job state</th>");
    result.push_str("<th>User name</th>");
    result.push_str("<th>User ID</th>");
    result.push_str("</tr>\n");

    for job in &status.job_info {
        result.push_str("<tr>\n");
        result.push_str(&format!("<td>{}</td>", job.executing_host));

        result.push_str(job.minimum_cpu.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.num_cpu.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.num_nodes.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.job_array_id.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.num_sockets.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.job_id.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.num_cores.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(&format!("<td>{}</td>", job.job_name));
        result.push_str(job.num_threads.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(job.job_array_index.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(&format!("<td>{}</td>", job.run_time));
        result.push_str("<td>");
        for node in &job.list_of_nodes {
            result.push_str(&format!("{},", node));
        }
        result.push_str("</td>");
        result.push_str(job.priority.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str(&format!("<td>{:?}</td>", job.state_reason));
        result.push_str(&format!("<td>{}</td>", job.start_time));
        result.push_str(&format!("<td>{:?}</td>", job.job_state));
        result.push_str(&format!("<td>{}</td>", job.user_name));
        result.push_str(job.user_id.map_or("<td>-</td>".to_string(), |val| format!("<td>{}</td>", val)).as_ref());
        result.push_str("</tr>\n");
    }

    result.push_str("</table>\n");

    result.push_str("</body>\n");
    result.push_str("</html>\n");

    result
}


src/request_handler.rs
//! The Iron web framework request handler and helper function

// System modules:
use std::sync::Mutex;

// External modules:
use iron::prelude::{Request, IronResult, Response};
use iron::headers::{Headers, ContentType};
use iron::mime::{Mime, TopLevel, SubLevel};
use iron::status;

// Internal modules:
use slurm_status::{status_to_html, SlurmStatus};

/// Accepts a HTML string and returns a IronResult response with correct mime type
fn string_to_response(page: &str) -> IronResult<Response> {
    let mut res = Response::new();

    res.status = Some(status::Ok);
    res.headers = Headers::new();
    res.headers.set(ContentType(Mime(TopLevel::Text, SubLevel::Html, vec![])));
    res.body = Some(Box::new(page.to_string()));

    Ok(res)
}

/// Handles Iron requests and shows the slurm status as a HTML web page
/// TODO: better error handling
pub fn handle_request(req: &mut Request, shared_slurm_status: &Mutex<SlurmStatus>) -> IronResult<Response> {
    debug!("req: {:?}", req);

    match shared_slurm_status.lock() {
        Ok(status) => {
            string_to_response(&status_to_html(&status))
        },
        Err(err) => {
            error!("Could not lock Mutex: {}", err);
            string_to_response("<h1>Could not lock Mutex!</h1>")
        }
    }
}


src/configuration.rs
//! Handles the configuration for slurm_inspector
//! Parses command line arguments via clap and sets default values

// External modules:
use clap::App;


/// slurm_inspector configuration (from command line arguments)
#[derive(Debug, Clone, PartialEq)]
pub struct Configuration {
    /// Iron webframework port, default: 4545
    pub port: u16,
    /// Update time intervall for SLURM status thread in seconds, default: 60 sec.
    pub interval: u64,
    /// If test mode is enabled, create some test data and do not call external commands ("sinfo", "squeue")
    pub test_mode: bool,
    /// Set the log level for flexi_logger: error, info or debug
    pub log_level: String
}

/// This will parse the command line arguments and create a new configuration object
/// If the arguments are missing or there is a parse error, then the default values are used
pub fn setup_configuration() -> Configuration {
    let matches = App::new("slurm_inspector")
        .version("0.1")
        .author("Willi Kappler")
        .about("Web GUI for the slurm job management system")
        .args_from_usage(
            "-p --port=[PORT] 'Sets the port for the web GUI (default: 4545)'
             -i --interval=[INTERVAL] 'Sets the update interval (in sec.) for the web page (default: 60 sec.)'
             --test 'create test values, does not call sinfo or squeue'
             --loglevel=[LOGLEVEL] 'specify log level: error, info or debug'"
        )
        .get_matches();

        let port = value_t!(matches.value_of("PORT"), u16).unwrap_or(4545);
        let interval = value_t!(matches.value_of("INTERVAL"), u64).unwrap_or(60);
        let test_mode = matches.is_present("test");
        let log_level = match matches.value_of("LOGLEVEL") {
            Some("error") => "error",
            Some("info") => "info",
            Some("debug") => "debug",
            _ => "info"
        };

        Configuration {
            port: port,
            interval: interval,
            test_mode: test_mode,
            log_level: log_level.to_string()
        }
}

#[test]
fn test_setup_configuration() {
    assert_eq!(setup_configuration(), Configuration{ port: 4545, interval: 60, test_mode: false, log_level: "info".to_string() });
}
